packages/eas-cli/src/utils/queries.ts

```diff
@@ -0,0 +1,138 @@
+import { confirmAsync, selectAsync } from '../prompts';
+import uniqBy from './expodash/uniqBy';
+
+const fetchMoreValue = '_fetchMore';
+
+export type PaginatedQueryResponse<QueryReturnType extends Record<string, any>> = {
+  queryResponse: QueryReturnType[];
+  queryResponseRawLength: number;
+};
+
+type BasePaginatedQueryArgs<QueryReturnType extends Record<string, any>> = {
+  pageSize: number;
+  offset: number;
+  queryToPerform: (
+    pageSize: number,
+    offset: number
+  ) => Promise<PaginatedQueryResponse<QueryReturnType>>;
+};
+
+type PaginatedQueryWithConfirmPromptArgs<QueryReturnType extends Record<string, any>> =
+  BasePaginatedQueryArgs<QueryReturnType> & {
+    promptOptions: {
+      readonly title: string;
+      renderListItems: (currentPage: QueryReturnType[]) => void;
+    };
+  };
+
+type PaginatedQueryWithSelectPromptArgs<QueryReturnType extends Record<string, any>> =
+  BasePaginatedQueryArgs<QueryReturnType> & {
+    promptOptions: {
+      readonly title: string;
+      createDisplayTextForSelectionPromptListItem: (queryItem: QueryReturnType) => string;
+      getIdentifierForQueryItem: (queryItem: QueryReturnType) => string;
+    };
+  };
+
+export async function paginatedQueryWithConfirmPromptAsync<
+  QueryReturnType extends Record<string, any>
+>(
+  {
+    pageSize,
+    offset,
+    queryToPerform,
+    promptOptions,
+  }: PaginatedQueryWithConfirmPromptArgs<QueryReturnType>,
+  accumulator: QueryReturnType[] = []
+): Promise<void> {
+  // query an extra item to determine if there are more pages left
+  const { queryResponse: items, queryResponseRawLength } = await queryToPerform(
```

Comment:
technically you don't need to handle an empty list here like in select because it won't block users, but UX here might not be ideal. In part that it's in this PR you are using a table so it would result in displaying just the header of the table which is probably fine, but there might be other cases in follow-up PRs that would not produce any output in case fo empty list.

Nothing actionable here, but keep that in mind for other parts.

Reply:
It seemed correct to not break the pagination logic when a user modifies the query results before passing them back. For example, here's an implementation that will be in a future PR:
```javascript
const queryUpdatesForBranchAsync = async (
    pageSize: number,
    offset: number
  ): Promise<PaginatedQueryResponse<AppUpdateObject>> => {
    const { updates } = await ensureBranchExistsAsync({
      appId: projectId,
      name: branchName!,
      limit: pageSize,
      offset,
    });
    if (!updates.length) {
      throw new Error(
        `There are no updates on branch "${branchName}" published for the platform(s) ${platformFlag}. Did you mean to publish a new update instead?`
      );
    }
    const relevantUpdates =
      platformFlag === 'all'
        ? updates
        : updates.filter(update => {
            // Only show groups that have updates on the specified platform(s).
            return update.platform === platformFlag;
          });
    return {
      queryResponse: relevantUpdates,
      queryResponseRawLength: updates.length,
    };
  };
  ```
  
  This would break the pagination, but we also don't want to return non-relevant updates for a user to view/select.
  
  There are definitely other ways we can approach this, like adding additional arguments to the select function... `omitItemFromSelectionList` but this implementation seemed to provide the most flexibility with the least verbosity which is why i went for it.
  
Do you have feedback? Please feel free to reply with some pseudo code or whatever.

Reply:
re:
> technically you don't need to handle an empty list here like in select because it won't block users, but UX here might not be ideal.

I'm currently handling it at the individual query helper level. i.e. `src/update/queries/queryUpdateGroupsForBranchAsync` will throw if no items are returned by the query.

Reply:
> My thinking is: it seems correct to support user modifications to the query results before passing them back. For example, here's an implementation that will be in a future PR:

I'm still not convinced. It seems you only need this interface because the (server) API is limiting. It's best to change the API. For filtering by platform, you could do this on the server-side.

Re:
```
  return {
    queryResponse: [updateBranch],
    queryResponseRawLength: updateBranch.updates.length,
  };
```

This also doesn't look like something that justifies `queryResponseRawLength`. `queryResponseRawLength` can be 
accessed on the call site.